File: app/src/main/java/com/apps/wow/droider/Widget/FeedWidget.java
package com.apps.wow.droider.Widget;

import com.apps.wow.droider.Article.ArticleActivity;
import com.apps.wow.droider.R;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.widget.RemoteViews;
import android.widget.Toast;

import java.util.Arrays;

import static android.content.ContentValues.TAG;
import static com.apps.wow.droider.Utils.Utils.EXTRA_ARTICLE_IMG_URL;
import static com.apps.wow.droider.Utils.Utils.EXTRA_ARTICLE_TITLE;
import static com.apps.wow.droider.Utils.Utils.EXTRA_ARTICLE_URL;
import static com.apps.wow.droider.Utils.Utils.EXTRA_SHORT_DESCRIPTION;

/**
 * Created by a.guzenko on 06.05.17.
 */

public class FeedWidget extends AppWidgetProvider {

    static final String LOG_TAG = "myLogs";

    final String ACTION_ON_CLICK = "com.apps.wow.droider.Widget.itemonclick";

    final static String ITEM_POSITION = "item_position";

    private PendingIntent newPending;

    @Override
    public void onEnabled(Context context) {
        super.onEnabled(context);
        Log.d(LOG_TAG, "onEnabled");
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        super.onUpdate(context, appWidgetManager, appWidgetIds);

        for (int appWidgetId : appWidgetIds) {
            RemoteViews rv = new RemoteViews(context.getPackageName(), R.layout.widget);
            setList(rv, context, appWidgetId);
            setListClick(rv, context);
            appWidgetManager.updateAppWidget(appWidgetId, rv);
            appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.last_news);
            Log.d(LOG_TAG, "onUpdate " + Arrays.toString(appWidgetIds));
        }
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        super.onDeleted(context, appWidgetIds);
        Log.d(LOG_TAG, "onDeleted " + Arrays.toString(appWidgetIds));
        update(context);
    }

    @Override
    public void onDisabled(Context context) {
        super.onDisabled(context);
        Log.d(LOG_TAG, "onDisabled");
    }

    public static void update(Context context) {
        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
        if (appWidgetManager != null) {
            Log.d(LOG_TAG, "update: ");
            ComponentName name = new ComponentName(context, FeedWidget.class);
            int[] appWidgetIds = appWidgetManager.getAppWidgetIds(name);
            appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetIds, R.id.last_news);
        }
    }

    void setList(RemoteViews rv, Context context, int appWidgetId) {
        Intent adapter = new Intent(context, ListViewWidgetService.class);
        adapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        rv.setRemoteAdapter(R.id.last_news, adapter);
    }

    void setListClick(RemoteViews rv, Context context) {
        Intent listClickIntent = new Intent(context, FeedWidget.class);
        listClickIntent.setAction(ACTION_ON_CLICK);
        PendingIntent listClickPIntent = PendingIntent.getBroadcast(context, 0, listClickIntent, 0);
        rv.setPendingIntentTemplate(R.id.last_news, listClickPIntent);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
        if (intent.getAction().equalsIgnoreCase(ACTION_ON_CLICK)) {
            int itemPos = intent.getIntExtra(ITEM_POSITION, -1);
            if (itemPos != -1) {
                try {
                    Intent articleIntent = new Intent(context, ArticleActivity.class);
                    articleIntent.putExtra(EXTRA_ARTICLE_TITLE,
                            intent.getStringExtra(EXTRA_ARTICLE_TITLE));
                    articleIntent
                            .putExtra(EXTRA_ARTICLE_URL, intent.getStringExtra(EXTRA_ARTICLE_URL));
                    articleIntent.putExtra(EXTRA_ARTICLE_IMG_URL,
                            intent.getStringExtra(EXTRA_ARTICLE_IMG_URL));
                    articleIntent.putExtra(EXTRA_SHORT_DESCRIPTION, intent.getStringExtra(EXTRA_SHORT_DESCRIPTION));
                    articleIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    context.startActivity(articleIntent);

                } catch (NullPointerException npe) {
                    // Ошибка происходит если пытаться отправить пикчу
                    // в статью. Сначала он выкидывал NullPointerException
                    // на article в ArticleActivity. Я закомментил
                    // после этого ничего не открывалось
                    Toast.makeText(context, "Произошла ошибка при открытии статьи!",
                            Toast.LENGTH_LONG).show();
                    Log.e(TAG, "onClick: Failed to open ArticleActivity!", npe.getCause());
                    npe.printStackTrace();
                }
            }
        }
    }
}

File: app/src/main/java/com/apps/wow/droider/Widget/FeedWidget.kt
package com.apps.wow.droider.Widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.ContentValues.TAG
import android.content.Context
import android.content.Intent
import android.util.Log
import android.widget.RemoteViews
import android.widget.Toast
import com.apps.wow.droider.Article.ArticleActivity
import com.apps.wow.droider.R
import com.apps.wow.droider.Utils.Utils
import java.util.*

/**
 * Created by a.guzenko on 06.05.17.
 */

class FeedWidget : AppWidgetProvider() {

    internal val ACTION_ON_CLICK = "com.apps.wow.droider.Widget.itemonclick"

    override fun onEnabled(context: Context) {
        super.onEnabled(context)
        Log.d(LOG_TAG, "onEnabled")
    }

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)

        for (appWidgetId in appWidgetIds) {
            val rv = RemoteViews(context.packageName, R.layout.widget)
            setList(rv, context, appWidgetId)
            setListClick(rv, context)
            appWidgetManager.updateAppWidget(appWidgetId, rv)
            appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.last_news)
            Log.d(LOG_TAG, "onUpdate " + Arrays.toString(appWidgetIds))
        }
    }

    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
        super.onDeleted(context, appWidgetIds)
        Log.d(LOG_TAG, "onDeleted " + Arrays.toString(appWidgetIds))
        update(context)
    }

    override fun onDisabled(context: Context) {
        super.onDisabled(context)
        Log.d(LOG_TAG, "onDisabled")
    }

    internal fun setList(rv: RemoteViews, context: Context, appWidgetId: Int) {
        val adapter = Intent(context, ListViewWidgetService::class.java)
        adapter.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
        rv.setRemoteAdapter(R.id.last_news, adapter)
    }

    internal fun setListClick(rv: RemoteViews, context: Context) {
        val listClickIntent = Intent(context, FeedWidget::class.java)
        listClickIntent.action = ACTION_ON_CLICK
        val listClickPIntent = PendingIntent.getBroadcast(context, 0, listClickIntent, 0)
        rv.setPendingIntentTemplate(R.id.last_news, listClickPIntent)
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        if (intent.action.equals(ACTION_ON_CLICK, ignoreCase = true)) {
            val itemPos = intent.getIntExtra(ITEM_POSITION, -1)
            if (itemPos != -1) {
                try {
                    val articleIntent = Intent(context, ArticleActivity::class.java)
                    articleIntent.putExtra(Utils.EXTRA_ARTICLE_TITLE,
                            intent.getStringExtra(Utils.EXTRA_ARTICLE_TITLE))
                    articleIntent
                            .putExtra(Utils.EXTRA_ARTICLE_URL, intent.getStringExtra(Utils.EXTRA_ARTICLE_URL))
                    articleIntent.putExtra(Utils.EXTRA_ARTICLE_IMG_URL,
                            intent.getStringExtra(Utils.EXTRA_ARTICLE_IMG_URL))
                    articleIntent.putExtra(Utils.EXTRA_SHORT_DESCRIPTION, intent.getStringExtra(Utils.EXTRA_SHORT_DESCRIPTION))
                    articleIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                    context.startActivity(articleIntent)

                } catch (npe: NullPointerException) {
                    // Ошибка происходит если пытаться отправить пикчу
                    // в статью. Сначала он выкидывал NullPointerException
                    // на article в ArticleActivity. Я закомментил
                    // после этого ничего не открывалось
                    Toast.makeText(context, "Произошла ошибка при открытии статьи!",
                            Toast.LENGTH_LONG).show()
                    Log.e(TAG, "onClick: Failed to open ArticleActivity!", npe.cause)
                    npe.printStackTrace()
                }

            }
        }
    }

    companion object {

        internal val LOG_TAG = "myLogs"

        internal val ITEM_POSITION = "item_position"

        fun update(context: Context) {
            val appWidgetManager = AppWidgetManager.getInstance(context)
            if (appWidgetManager != null) {
                Log.d(LOG_TAG, "update: ")
                val name = ComponentName(context, FeedWidget::class.java)
                val appWidgetIds = appWidgetManager.getAppWidgetIds(name)
                appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetIds, R.id.last_news)
            }
        }
    }
}
